Introduce a switch between Merge and Apply Status Admission Patch

mszadkow opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
We want a feature gate for the ability to alternate between Merge and Apply for patching a status of the workload.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Relates to #6158

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note

```

mimowo commented on pkg/workload/workload.go:
```diff
@@ -857,6 +857,25 @@ func ApplyAdmissionStatusPatch(ctx context.Context, c client.Client, patch *kueu
+// PatchAdmissionStatusWithOriginalWorkload updates the admission status of a workload.
+// If the WorkloadRequestUseMergePatch feature is enabled, it uses a Merge Patch with update function.
+// Otherwise, it runs the update function and, if updated, applies the SSA Patch status.
+func PatchAdmissionStatusWithOriginalWorkload(ctx context.Context, c client.Client, wOrig *kueue.Workload, w *kueue.Workload, strict bool, clk clock.Clock, update func() (bool, error)) error {
```
Why we need this function? I would prefer to inline inside `PatchAdmissionStatus`.

mimowo commented on pkg/util/client/client.go:
```diff
@@ -35,7 +36,7 @@ func createPatch(before, after client.Object) (client.Patch, error) {
 // client.Object was changed.
 func Patch(ctx context.Context, c client.Client, obj client.Object, strict bool, update func() (bool, error)) error {
-	return updateAndPatch(obj, strict, update, func(patch client.Patch) error {
+	return updateAndPatch(nil, obj, strict, update, func(patch client.Patch) error {
```
I'm puzzled what is the purpose of the changes here

mszadkow replied:
Do you ask why the original obj is needed to make Patch with Merge?
If we compare the code to SSA it's way different creation of the patch.
With merge patch all of the changes has to be included in the `update()` or they will be not the part of the patch.
We have some very big functions that change many things in the workload and only then the workload is send to be patched.
Without providing the reference point for the workload, we would have to refactor those functions and it is risky.

mimowo commented on pkg/workload/workload.go:
```diff
@@ -857,6 +857,20 @@ func ApplyAdmissionStatusPatch(ctx context.Context, c client.Client, patch *kueu
+// PatchAdmissionStatus updates the admission status of a workload.
+// If the WorkloadRequestUseMergePatch feature is enabled, it uses a Merge Patch with update function.
+// Otherwise, it runs the update function and, if updated, applies the SSA Patch status.
+func PatchAdmissionStatus(ctx context.Context, c client.Client, wOrig *kueue.Workload, w *kueue.Workload, strict bool, clk clock.Clock, update func() (bool, error)) error {
```
This API seems very weird. I think you should only need to accept wOrig  + update.

mbobrovskyi commented on pkg/workload/workload.go:
```diff
@@ -1062,14 +1080,53 @@ func AdmissionChecksForWorkload(log logr.Logger, wl *kueue.Workload, admissionCh
-		evictionReason = fmt.Sprintf("%sDueTo%s", evictionReason, underlyingCause)
+type EvictionOption func(*EvictionOptions)
+
+type EvictionOptions struct {
```
```suggestion
type EvictOptions struct {
```
nit

mbobrovskyi commented on pkg/workload/workload.go:
```diff
@@ -1062,14 +1080,53 @@ func AdmissionChecksForWorkload(log logr.Logger, wl *kueue.Workload, admissionCh
+	}
+}
+
+func Evict(ctx context.Context, c client.Client, recorder record.EventRecorder, wlOrig *kueue.Workload, reason, underlyingCause, msg string, clock clock.Clock, options ...EvictionOption) error {
```
The `underlyingCause` argument also not required. I'm wondering if we can move it to EvictOptions as well. It can be implemented in follow-up PR.

mbobrovskyi commented on pkg/controller/jobframework/reconciler.go:
```diff
@@ -535,10 +535,13 @@ func (r *JobReconciler) ReconcileGenericJob(ctx context.Context, req ctrl.Reques
 				log.V(6).Info("The job is no longer active, clear the workloads admission")
 				// The requeued condition status set to true only on EvictedByPreemption
+				wlOrig := wl.DeepCopy()
 				setRequeued := evCond.Reason == kueue.WorkloadEvictedByPreemption
 				workload.SetRequeuedCondition(wl, evCond.Reason, evCond.Message, setRequeued)
 				_ = workload.UnsetQuotaReservationWithCondition(wl, "Pending", evCond.Message, r.clock.Now())
```
I'm wondering if we can move it inside customPrepare function and check if it is changes.

mbobrovskyi commented on pkg/scheduler/preemption/preemption.go:
```diff
@@ -193,10 +193,13 @@ func (p *Preemptor) IssuePreemptions(ctx context.Context, preemptor *workload.In
 
-func (p *Preemptor) applyPreemptionWithSSA(ctx context.Context, w *kueue.Workload, reason, message string) error {
+func (p *Preemptor) patchPreemption(ctx context.Context, w *kueue.Workload, reason, message string) error {
+	wOrig := w.DeepCopy()
 	w = w.DeepCopy()
```
Do we need to duplicate this object twice? Wouldn't one copy be sufficient?

mbobrovskyi commented on pkg/workload/workload.go:
```diff
@@ -1062,14 +1080,53 @@ func AdmissionChecksForWorkload(log logr.Logger, wl *kueue.Workload, admissionCh
+	}
+
+	// if there is no customPrepare use wl and wlOrig are equal
+	wl := wlOrig
```
```suggestion
	wl := wlOrig.DeepCopy()
```

mszadkow commented on pkg/scheduler/scheduler.go:
```diff
@@ -764,17 +770,19 @@ func (s *Scheduler) requeueAndUpdate(ctx context.Context, e entry) {
-			}
+		wl := e.Obj.DeepCopy()
+		if err := workload.PatchAdmissionStatus(ctx, s.client, wlOrig, true, s.clock, func() (*kueue.Workload, bool, error) {
+			workload.SetRequeuedCondition(wl, kueue.WorkloadClusterQueueRestarted, "The ClusterQueue was restarted after being stopped", true)
```
but it doesn't mean I will not check it

mbobrovskyi replied:
What about SetRequeuedCondition changes?

mszadkow replied:
oh...that's unexpected...

mszadkow commented on pkg/scheduler/preemption/preemption.go:
```diff
@@ -193,10 +193,13 @@ func (p *Preemptor) IssuePreemptions(ctx context.Context, preemptor *workload.In
+func (p *Preemptor) patchPreemption(ctx context.Context, w *kueue.Workload, reason, message string) error {
+	wOrig := w.DeepCopy()
 	w = w.DeepCopy()
 	workload.SetPreemptedCondition(w, reason, message)
```
Done.

ichekrygin commented on pkg/scheduler/scheduler.go:
```diff
@@ -84,7 +84,7 @@ type Scheduler struct {
 
 	// Stubs.
-	applyAdmission func(context.Context, *kueue.Workload) error
+	patchAdmission func(context.Context, *kueue.Workload, *kueue.Workload) error
```
Small nit: consider adding arg names to make it clear which workload is which, i.e. (or similar):
```suggestion
	patchAdmission func(ctx context.Context, original *kueue.Workload, updated *kueue.Workload) error
```

mszadkow replied:
Or maybe I will do that as the prep PR, as reviewing this one might take a while.

mszadkow commented on pkg/controller/admissionchecks/multikueue/workload.go:
```diff
@@ -340,14 +340,13 @@ func (w *wlReconciler) reconcileGroup(ctx context.Context, group *wlGroup) (reco
 			Status:  metav1.ConditionTrue,
 			Reason:  remoteFinishedCond.Reason,
 			Message: remoteFinishedCond.Message,
 		})
```
I think it's reasonable to align here.

mszadkow commented on pkg/workload/workload.go:
```diff
@@ -1103,17 +1121,54 @@ func AdmissionChecksForWorkload(log logr.Logger, wl *kueue.Workload, admissionCh
+	}
+}
+
+func Evict(ctx context.Context, c client.Client, recorder record.EventRecorder, wlOrig *kueue.Workload, reason, underlyingCause, msg string, clock clock.Clock, options ...EvictOption) error {
```
Very interesting point, I would go for that in this particular case, it indeed simplifies all here.

mbobrovskyi commented on pkg/controller/core/workload_controller.go:
```diff
@@ -517,14 +549,18 @@ func (r *WorkloadReconciler) reconcileOnLocalQueueActiveState(ctx context.Contex
-		_ = workload.UnsetQuotaReservationWithCondition(wl, kueue.WorkloadInadmissible, fmt.Sprintf("LocalQueue %s is terminating or missing", wl.Spec.QueueName), r.clock.Now())
-		return true, workload.ApplyAdmissionStatus(ctx, r.client, wl, true, r.clock)
+		return true, workload.PatchAdmissionStatus(ctx, r.client, wl, true, r.clock, func() (*kueue.Workload, bool, error) {
+			_ = workload.UnsetQuotaReservationWithCondition(wl, kueue.WorkloadInadmissible, fmt.Sprintf("LocalQueue %s is terminating or missing", wl.Spec.QueueName), r.clock.Now())
+			return wl, true, nil
```
```suggestion
			return wl, workload.UnsetQuotaReservationWithCondition(wl, kueue.WorkloadInadmissible, fmt.Sprintf("LocalQueue %s is terminating or missing", wl.Spec.QueueName), r.clock.Now()), nil
```
Maybe like this?

mbobrovskyi commented on pkg/controller/jobs/job/job_controller_test.go:
```diff
@@ -586,6 +586,7 @@ func TestReconciler(t *testing.T) {
 		enableObjectRetentionPolicies                     bool
 		enableTopologyAwareScheduling                     bool
 		enableManagedJobsNamespaceSelectorAlwaysRespected bool
+		removeAdmissionAsInRealCluster                    bool
```
Oh, I think this needs some explanation.

mbobrovskyi commented on pkg/controller/jobs/pod/pod_controller_test.go:
```diff
@@ -5579,85 +5582,96 @@ func TestReconciler(t *testing.T) {
-						t.Fatalf("Could not delete workload: %v", err)
+				kClient := kcBuilder.Build()
+				for i := range tc.workloads {
+					tc.workloads[i].ResourceVersion = ""
```
Why this?

mbobrovskyi commented on pkg/controller/jobs/job/job_controller_test.go:
```diff
@@ -3853,84 +3865,92 @@ func TestReconciler(t *testing.T) {
-				t.Errorf("Workloads after reconcile (-want,+got):\n%s", diff)
-			}
+				if features.Enabled(features.WorkloadRequestUseMergePatch) && tc.removeAdmissionAsInRealCluster {
+					tc.wantWorkloads[0].Status.Admission = nil
```
I’m wondering if we can just ignore this field when the feature is disabled, since this is a bug with Apply Patch.

mbobrovskyi commented on pkg/features/kube_features.go:
```diff
@@ -304,6 +309,9 @@ var defaultVersionedFeatureGates = map[featuregate.Feature]featuregate.Versioned
 		{Version: version.MustParse("0.14"), Default: false, PreRelease: featuregate.Alpha},
 	},
+	WorkloadRequestUseMergePatch: {
+		{Version: version.MustParse("0.14"), Default: true, PreRelease: featuregate.Alpha},
```
```suggestion
		{Version: version.MustParse("0.14"), Default: false, PreRelease: featuregate.Alpha},
```
Don't forget to revert it.

mimowo reviewed (commented):
I'm a bit concerned in this PR about possible performance regression by taking the DeepCopy() of workload even if we don't need it. It seems is used for the CustomPrepare. I'm wondering it should not be necessary if the Patch engine called DeepCopy, and then call CustomPrepare on the copy.

mimowo commented:
> I will try to minimise copying as much as possible or maybe try to slightly change the approach with passing current workload in the update/ costume prepare

Yes, I think basically we should only pass origWorkload, then both "Prepare" and "Update" should accept the "current" workload and return "new". Basically run "sztafeta" (polish word), sorry I don't know English word :) )

mimowo commented:
Regarding https://github.com/kubernetes-sigs/kueue/pull/6765#issuecomment-3328358698 I imagine another prep PR which does noting more than just enables this main PR to introduce the switch easily.

mszadkow commented:
> Regarding [#6765 (comment)](https://github.com/kubernetes-sigs/kueue/pull/6765#issuecomment-3328358698) I imagine another prep PR which does noting more than just enables this main PR to introduce the switch easily.

I am testing the idea and yes, if that works I think it would be great to have it as a prep

mbobrovskyi commented on pkg/controller/admissionchecks/multikueue/workload.go:
```diff
@@ -340,12 +342,26 @@ func (w *wlReconciler) reconcileGroup(ctx context.Context, group *wlGroup) (reco
-			Type:    kueue.WorkloadFinished,
-			Status:  metav1.ConditionTrue,
-			Reason:  remoteFinishedCond.Reason,
-			Message: remoteFinishedCond.Message,
+			Type:               kueue.WorkloadFinished,
+			Status:             metav1.ConditionTrue,
+			Reason:             remoteFinishedCond.Reason,
+			Message:            remoteFinishedCond.Message,
+			LastTransitionTime: metav1.NewTime(w.clock.Now()),
 		})
```
This looks the same as the merge patch. Can we generalize it?

mbobrovskyi commented on pkg/controller/core/workload_controller_test.go:
```diff
@@ -2252,151 +2438,164 @@ func TestReconcile(t *testing.T) {
-			features.SetFeatureGateDuringTest(t, features.ObjectRetentionPolicies, tc.enableObjectRetentionPolicies)
-			features.SetFeatureGateDuringTest(t, features.DynamicResourceAllocation, tc.enableDRAFeature)
+		for _, enabled := range []bool{false, true} {
+			tc := tc
```
What is it?

mbobrovskyi commented on pkg/controller/core/workload_controller_test.go:
```diff
@@ -2252,151 +2438,164 @@ func TestReconcile(t *testing.T) {
+						}
+						t.Fatalf("expected workload to persist")
+					}
+					if features.Enabled(features.WorkloadRequestUseMergePatch) && tc.wantWorkloadUseMergePatch != nil {
```
Is it the changes only with Admission field in Status? I'm wondering if we can just ignore it for Apply Patch using cmpopts.IgnoreField?

mszadkow replied:
Nope, here is more complicated, not as easy as the other ones

mbobrovskyi commented on pkg/controller/jobs/job/job_controller_test.go:
```diff
@@ -533,6 +534,9 @@ var (
 		),
 		cmpopts.IgnoreFields(metav1.Condition{}, "LastTransitionTime"),
 		cmpopts.IgnoreFields(kueue.AdmissionCheckState{}, "LastTransitionTime"),
+		// Fake client with patch.Apply can't reset Admission field, patch.Merge can
+		// However other key Status fields indicate that change e.g. Conditions, thuse we choose to ignore the Admission field
+		cmpopts.IgnoreFields(kueue.WorkloadStatus{}, "Admission"),
```
Please add it only for ApplyPatch and only if it is really need.

mbobrovskyi commented on pkg/controller/jobs/job/job_controller_test.go:
```diff
@@ -3853,84 +3857,89 @@ func TestReconciler(t *testing.T) {
-			features.SetFeatureGateDuringTest(t, features.ObjectRetentionPolicies, tc.enableObjectRetentionPolicies)
-			features.SetFeatureGateDuringTest(t, features.ManagedJobsNamespaceSelectorAlwaysRespected, tc.enableManagedJobsNamespaceSelectorAlwaysRespected)
+		for _, enabled := range []bool{false, true} {
+			tc := tc
```
```suggestion
```

mbobrovskyi commented on pkg/controller/jobs/job/job_controller_test.go:
```diff
@@ -3853,84 +3857,89 @@ func TestReconciler(t *testing.T) {
+							t.Fatalf("Could not setup owner reference in Workloads: %v", err)
+						}
+					}
+					tc.workloads[i].ResourceVersion = ""
```
Try to duplicate workload instead.

mbobrovskyi commented on pkg/scheduler/scheduler.go:
```diff
@@ -565,7 +567,7 @@ func (s *Scheduler) getInitialAssignments(log logr.Logger, wl *workload.Info, sn
 	log.V(3).Info("Evicting workload after failed try to find a node replacement; TASFailedNodeReplacementFailFast enabled")
 	msg := fmt.Sprintf("Workload was evicted as there was no replacement for a failed node: %s", wl.Status.UnhealthyNodes[0].Name)
-	if err := workload.Evict(ctx, s.client, s.recorder, wl, kueue.WorkloadEvictedDueToNodeFailures, msg, "", s.clock); err != nil {
+	if err := workload.Evict(ctx, s.client, s.recorder, wl.DeepCopy(), kueue.WorkloadEvictedDueToNodeFailures, msg, "", s.clock); err != nil {
```
We should DeepCopy it inside Evict function.

mszadkow replied:
This DeepCopy must stay as otherwise we have race condition, due to how PatchStatus works.

mbobrovskyi commented on pkg/scheduler/scheduler.go:
```diff
@@ -771,7 +774,8 @@ func (s *Scheduler) requeueAndUpdate(ctx context.Context, e entry) {
 	log.V(2).Info("Workload re-queued", "workload", klog.KObj(e.Obj), "clusterQueue", klog.KRef("", string(e.ClusterQueue)), "queue", klog.KRef(e.Obj.Namespace, string(e.Obj.Spec.QueueName)), "requeueReason", e.requeueReason, "added", added, "status", e.status)
 	if e.status == notNominated || e.status == skipped {
-		if err := workload.PatchAdmissionStatus(ctx, s.client, e.Obj, true, s.clock, func() (*kueue.Workload, bool, error) {
+		wlOrig := e.Obj.DeepCopy()
```
```suggestion
```

mbobrovskyi commented on pkg/controller/jobs/job/job_controller_test.go:
```diff
@@ -534,6 +535,7 @@ var (
 		cmpopts.IgnoreFields(metav1.Condition{}, "LastTransitionTime"),
 		cmpopts.IgnoreFields(kueue.AdmissionCheckState{}, "LastTransitionTime"),
 	}
+
```
And this

mbobrovskyi commented:
/lgtm

Thank you so much!

mszadkow commented:
> > > Yes, I think basically we should only pass origWorkload, then both "Prepare" and "Update" should accept the "current" workload and return "new". Basically run "sztafeta" (polish word), sorry I don't know English word :) )
> > 
> > 
> > @mimowo Unfortunately, we can't implement this change in the current PR. The main issue is that in some cases, we can't update the workload directly within the callback function. A key example is [here](https://github.com/epam/kubernetes-kueue/blob/ee84b39ae4743ec45cb928003633258823b65861/pkg/scheduler/scheduler.go#L606-L644), where the workload is updated before the goroutine is executed, which poses a problem. Fixing this would require significant refactoring, which is beyond the scope of this PR.
> > I suggest addressing this in a follow-up PR, alongside implementing support for [issue #6992](https://github.com/kubernetes-sigs/kueue/issues/6992).
> 
> I see thank you for the summary and review comments. The PR looks much better than before. Still scheduler is performance sensitive so taking DeepCopy when the feature gate is disabled means a bit of a penalty hit, so I hope there is a way to refactor in the future that this is not necessary. In particular I would consider this can be done in Prepare.
> 
> I will make a decision today. Probable it could go as is

I think there is a way and it's rather connected to how we do the patch.
We can create another version of PatchStatus:
```go
func PatchStatus(ctx context.Context, c client.Client, obj client.Object, update func() (client.Object, bool, error), options ...PatchOption) error {
	return patchCommon(obj, update, func(objOrig client.Object, patch client.Patch) error {
		return c.Status().Patch(ctx, obj, patch)
	}, options...)
}


func updateAndPatch(obj client.Object, options PatchOptions, update func() (client.Object, bool, error), patchFn func(client.Object, client.Patch) error) error {
	objOriginal := getOriginalObject(obj, options)
	objPatched, updated, err := update()
	if err != nil || !updated {
		return err
	}
	patch, err := createPatch(objOriginal, objPatched)
	if err != nil {
		return err
	}
	return patchFn(objOriginal, patch)
}
```
Right now we rely on the PatchStatus which returns the object after update, with this adjustment we use object that we already copied for the patch.
I believe we can reduce number of deep copies in scheduler with that, bc most of the extra copies are used to avoid race conditions.

mbobrovskyi commented:
> I meant to have another version of PatchStatus for this purpose
Maybe use new Option for this.

mimowo reviewed (commented):
/lgtm
/approve
Thank you for the effort @mszadkow @mbobrovskyi 👍 

I know making an abstraction layer to switch between Apply and Patch is not so easy, especially in scheduler where the Apply construction and execution happens in different goroutines. I believe these challenges are solvable, and there is a room to reduce the number of DeepCopies. 

However:
1. We have performance tests framework and if the regression was signifficant, then it would show up, but it keeps passing on this PR. 
2.  I'm not totally sure what is the best way to structure the code at the moment. I would very much look forward to some follow up proposals.

cc @mwysokin

mimowo commented:
/lgtm
